crates/sui-protocol-config-macros/src/lib.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::{parse_macro_input, Data, DeriveInput, Fields, Type};
+
+/// This macro derives getters for each field in ProtocolConfig at a specific version
+/// If the field is empty, None is returned, otherwise Some(value) is returned
+#[proc_macro_derive(ProtocolConfigGetters)]
+pub fn getters_macro(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+
+    let struct_name = &ast.ident;
+    let data = &ast.data;
+
+    let getters = match data {
+        // For each field, get the name and type
+        Data::Struct(data_struct) => match &data_struct.fields {
+            Fields::Named(fields_named) => fields_named.named.iter().filter_map(|field| {
+                let field_name = field.ident.as_ref().expect("Field must be named");
+                let field_type = &field.ty;
+                match field_type {
+                    Type::Path(type_path)
+                        // Only derive for fields of type Option<T>
+                        if type_path
+                            .path
+                            .segments
+                            .last()
+                            .map_or(false, |segment| segment.ident == "Option") =>
+                    {
+                        // Prepend the function name with "get_for_current_version_"
```

Comment:
let's not do this - it seems to invert the usual pattern. if you have a `&ProtocolConfig` that is a reference to a protocol config for a given version. Here you're generating a method looks like

```
fn get_foo(&self) -> T {
  ProtocolConfig::get_for_version_impl(self.version).foo
}
```

When we should just be doing:

```
fn get_foo(&self) -> T {
  self.foo
}
```

`get_for_version_impl` is not a cheap method and we definitely don't want to be calling it in the execution path. 99.9% of the time we will simply be using the pre-constructed ProtocolConfig that lives in the epoch store.

Additionally we don't need separate methods for getting for a particular version, because you can already do that by first getting the config for a given version, and then using an accessor.

Lastly, note that we already have getters for all the existing methods, so I really don't think it is a big deal to add a few more instead of using a proc macro. But if you really want to add a proc macro, please delete the existing getters and generate them via the proc macro, instead of leaving them in place.

Reply:
I have some thoughts on directions we could take.
But for now I've simplified this PR so as not to conflate the core change here with that of broader protocol config enhancements. 
I've removed the proc macro and only added a getter for the new fields.
In a subsequent PR, I'll put out some protocol config proposals.